torchrl/modules/models/model_based.py

```diff
@@ -0,0 +1,292 @@
+            self.rssm_posterior(_tensordict)
+
+            tensordict_out.append(_tensordict)
+            if t < time_steps - 1:
+                _tensordict = step_mdp(
+                    _tensordict.select(*self.out_keys), keep_other=False
+                )
+                _tensordict = update_values[..., t + 1].update(_tensordict)
+
+        return torch.stack(tensordict_out, tensordict.ndimension() - 1).contiguous()
+
+
+class RSSMPrior(nn.Module):
+    """The prior network of the RSSM.
+
+    This network takes as input the previous state and belief and the current action.
+    It outputs the next prior state and belief, as well as the parameters of the prior state distribution.
+    State is by construction stochastic and belief is deterministic. In the "Dream to control" paper, these are called "deterministic state " and "stochastic state", respectively.
+    We distinguish states sampled according prior and posterior distribution for clarity.
+
+    Reference: https://arxiv.org/abs/1811.04551
+
+    Args:
+        hidden_dim (int, optional): Number of hidden units in the linear network. Input size of the recurrent network.
+        rnn_hidden_dim (int, optional): Number of hidden units in the recurrent network. Also size of the belief.
+        state_dim (int, optional): Size of the state.
+        action_spec (TensorSpec, optional): Action spec. If None an error will be raised when initializing.
+
+    """
+
+    def __init__(
+        self, hidden_dim=200, rnn_hidden_dim=200, state_dim=30, action_spec=None
+    ):
+        super().__init__()
+
+        # Prior
+        self.rnn = nn.GRUCell(hidden_dim, rnn_hidden_dim)
+        self.action_state_projector = nn.Sequential(nn.LazyLinear(hidden_dim), nn.ELU())
+        self.rnn_to_prior_projector = NormalParamWrapper(
+            nn.Sequential(
+                nn.Linear(hidden_dim, hidden_dim),
+                nn.ELU(),
+                nn.Linear(hidden_dim, 2 * state_dim),
+            ),
+            scale_lb=0.1,
+            scale_mapping="softplus",
+        )
+
+        self.state_dim = state_dim
+        self.rnn_hidden_dim = rnn_hidden_dim
+
+        if action_spec is None:
+            raise ValueError("action_spec must be provided")
+        self.action_shape = action_spec.shape
+
+    def forward(self, state, belief, action):
+        projector_input = torch.cat([state, action], dim=-1)
+        action_state = self.action_state_projector(projector_input)
+        belief = self.rnn(action_state, belief)
+        prior_mean, prior_std = self.rnn_to_prior_projector(belief)
+        state = prior_mean + torch.randn_like(prior_std) * prior_std
+        return prior_mean, prior_std, state, belief
+
+
+class RSSMPosterior(nn.Module):
+    """The posterior network of the RSSM.
+
+    This network takes as input the belief and the associated encoded observation.
+    It outputs the next posterior state and belief, as well as the parameters of the posterior state distribution.
```

Comment:
> It outputs the next posterior state and belief

not really, there is no belief. Also a more appropriate way of saying this would be "It returns the parameters of the posterior as well as a state sampled according to this distribution"